File: tools/axiom/axiom_mrurecent_parser.py
import os
import pandas as pd
from utils.discovery import find_artifact_files, load_csv_with_progress
from collector.collector import add_rows
from utils.logger import print_and_log

def process_axiom_mrurecent(axiom_dir: str, batch_size: int, base_dir: str):
    artifact_name = "Axiom_MRURecent"
    print_and_log(f"[{artifact_name}] Scanning for relevant CSVs under: {axiom_dir}")

    recent_files = find_artifact_files(axiom_dir, base_dir, artifact_name)
    if not recent_files:
        print_and_log(f"[{artifact_name}] No MRU Recent files found.")
        return

    for file_path in recent_files:
        print_and_log(f"[{artifact_name}] Processing: {file_path}")
        total_rows = 0
        try:
            for df in load_csv_with_progress(file_path, batch_size, artifact_name=artifact_name):
                timeline_data = []

                for _, row in df.iterrows():
                    timestamp = row.get("Registry Key Modified Date/Time - UTC+00:00 (M/d/yyyy)", "")
                    dt = pd.to_datetime(timestamp, utc=True, errors="coerce")
                    if pd.isnull(dt):
                        continue

                    data_path = row.get("File/Folder Link", "")
                    data_path = data_path if isinstance(data_path, str) else ""

                    if data_path:
                        if os.path.splitext(data_path)[1]:
                            data_details = os.path.basename(data_path)
                        else:
                            data_details = os.path.basename(os.path.normpath(data_path))
                    else:
                        data_details = row.get("File/Folder Name", "")

                    timeline_row = {
                        "DateTime": dt.isoformat().replace("+00:00", "Z"),
                        "Tool": "Axiom",
                        "ArtifactName": "MRU_Recent",
                        "TimestampInfo": "Key Last Modified",
                        "Description": "Shortcut Accessed",
                        "DataPath": data_path,
                        "DataDetails": data_details,
                        "EvidencePath": os.path.relpath(file_path, base_dir) if base_dir else file_path
                    }

                    timeline_data.append(timeline_row)

                total_rows += len(timeline_data)
                add_rows(timeline_data)
        except Exception as e:
            print_and_log(f"[{artifact_name}] Failed to parse {file_path}: {e}")
            continue

        print_and_log(f"[✓] Parsed {total_rows} timeline rows from: {file_path}")

File: ForensicTimeliner.Core/Tools/Axiom/AxiomMruRecentParser.cs
using CsvHelper;
using CsvHelper.Configuration;
using ForensicTimeliner.Models;
using ForensicTimeliner.Interfaces;
using ForensicTimeliner.Models;
using ForensicTimeliner.Utils;
using Spectre.Console;
using System.Globalization;

namespace ForensicTimeliner.Tools.Axiom;

public class AxiomMruRecentParser : IArtifactParser
{
    public List<TimelineRow> Parse(string inputDir, string baseDir, ArtifactDefinition artifact, ParsedArgs args)
    {
        var rows = new List<TimelineRow>();

        Logger.PrintAndLog($"[>] - [{artifact.Artifact}] Scanning for relevant CSVs under: [{inputDir}]", "SCAN");

        var files = Discovery.FindArtifactFiles(inputDir, baseDir, artifact.Artifact);
        if (!files.Any())
        {
            Logger.PrintAndLog($"[!] - [{artifact.Artifact}] No matching files found in: {inputDir}", "WARN");
            return rows;
        }

        foreach (var file in files)
        {
            int timelineCount = 0;
            Logger.PrintAndLog($"[+] - [{artifact.Artifact}] Processing: {Path.GetRelativePath(baseDir, file)}", "PROCESS");

            try
            {
                using var reader = new StreamReader(file);
                using var csv = new CsvReader(reader, new CsvConfiguration(CultureInfo.InvariantCulture)
                {
                    HeaderValidated = null,
                    MissingFieldFound = null
                });

                var records = csv.GetRecords<dynamic>();
                foreach (var record in records)
                {
                    var dict = (IDictionary<string, object>)record;

                    var parsedDt = dict.GetDateTime("Registry Key Modified Date/Time - UTC+00:00 (M/d/yyyy)");
                    if (parsedDt == null) continue;

                    string dataPath = dict.GetString("File/Folder Link") ?? "";
                    string dataDetails;

                    if (!string.IsNullOrWhiteSpace(dataPath))
                    {
                        dataDetails = Path.HasExtension(dataPath)
                            ? Path.GetFileName(dataPath)
                            : Path.GetFileName(Path.GetDirectoryName(dataPath) ?? "") ?? "";
                    }
                    else
                    {
                        dataDetails = dict.GetString("File/Folder Name") ?? "";
                    }

                    string dtStr = parsedDt.Value.ToString("o").Replace("+00:00", "Z");


                    rows.Add(new TimelineRow
                    {
                        DateTime = dtStr,
                        TimestampInfo = "Last Modified",
                        ArtifactName = "Registry",
                        Tool = artifact.Tool,
                        Description = "MRU Recent",
                        DataPath = dataPath,
                        DataDetails = dataDetails,
                        EvidencePath = Path.GetRelativePath(baseDir, file)
                    });

                    timelineCount++;
                }

                Logger.PrintAndLog($"[✓] - [{artifact.Artifact}] Parsed {timelineCount} timeline rows from: {Path.GetFileName(file)}", "SUCCESS");
                LoggerSummary.TrackSummary(artifact.Tool, artifact.Artifact, timelineCount);
            }
            catch (Exception ex)
            {
                Logger.PrintAndLog($"[{artifact.Artifact}] Failed to parse {file}: {ex.Message}", "ERROR");
            }
        }

        return rows;
    }
}

